File: app/frontend/src/types/finetune.ts
import type { ColumnMapping } from "./dataset";
import type { ForecastResponse } from "./forecast";

export interface FinetuneConfig {
  epochs: number;
  batch_size: number;
  lr: number;
  context_len: number;
  horizon_len: number;
  lora_r: number;
  lora_alpha: number;
  lora_dropout: number;
  num_samples: number;
  seed: number;
}

export interface FinetuneRequest {
  dataset_id: string;
  mapping: ColumnMapping;
  config: FinetuneConfig;
  adapter_name: string;
}

export interface JobCreated {
  job_id: string;
}

export interface JobStatus {
  job_id: string;
  status: "running" | "done" | "error" | "cancelled";
  error?: string | null;
  adapter_id?: string | null;
}

export interface AdapterMetadata {
  adapter_id: string;
  name: string;
  created_at: string;
  dataset_id: string;
  best_val_loss: number | null;
  config: FinetuneConfig;
}

export interface AdapterForecastRequest {
  adapter_id: string;
  dataset_id: string;
  mapping: ColumnMapping;
  horizon: number;
}

export type { ForecastResponse };

export interface ProgressEvent {
  type: "progress";
  epoch: number;
  total_epochs: number;
  batch: number;
  total_batches: number;
  train_loss: number;
  timestamp: number;
}

export interface EpochEvent {
  type: "epoch";
  epoch: number;
  total_epochs: number;
  train_loss: number;
  val_loss: number;
  best: boolean;
  timestamp: number;
}

export interface LogEvent {
  type: "log";
  message: string;
  timestamp: number;
}

export interface DoneEvent {
  type: "done";
  adapter_id: string;
  best_val_loss: number;
  timestamp: number;
}

export interface ErrorEvent {
  type: "error";
  message: string;
  timestamp: number;
}

export type JobStatusEvent =
  | ProgressEvent
  | EpochEvent
  | LogEvent
  | DoneEvent
  | ErrorEvent;

File: app/backend/foreko/schemas/finetune.py
"""Fine-tuning request, response, and adapter metadata schemas."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from .dataset import ColumnMapping
from .jobs import JobCreated, JobStatus  # re-export

__all__ = ["FinetuneConfig", "FinetuneRequest", "AdapterMetadata", "AdapterForecastRequest", "JobCreated", "JobStatus"]


class FinetuneConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    epochs: int = Field(default=5, ge=1, le=100)
    batch_size: int = Field(default=32, ge=1, le=256)
    lr: float = Field(default=1e-4, gt=0)
    context_len: int = Field(default=64, ge=32, le=512)
    horizon_len: int = Field(default=13, ge=1, le=512)
    lora_r: int = Field(default=4, ge=1, le=64)
    lora_alpha: int = Field(default=8, ge=1)
    lora_dropout: float = Field(default=0.05, ge=0.0, le=0.5)
    num_samples: int = Field(default=2000, ge=10, le=50000)
    seed: int = 42


class FinetuneRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    dataset_id: str
    mapping: ColumnMapping
    config: FinetuneConfig = Field(default_factory=FinetuneConfig)
    adapter_name: str = ""


class AdapterMetadata(BaseModel):
    adapter_id: str
    name: str
    created_at: str
    dataset_id: str
    best_val_loss: float | None = None
    config: FinetuneConfig


class AdapterForecastRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    adapter_id: str
    dataset_id: str
    mapping: ColumnMapping
    horizon: int = Field(default=12, ge=1, le=512)
    context_len: int = Field(default=512, ge=32, le=512)
